### PR TITLE
Folia Compatibility Utility Rewrite

### PR DESCRIPTION
**Pull request by Bram1903**
# Folia Compatibility Utility Rewrite

This PR is focused on improving the current utility class with an improved version. The old one seems to be implemented in a hurry and hasn't been given much thought. It only supports a few basic scheduler-specific methods, and therefore I thought I'll give it a shot, and see if I am capable of improving it. 

## Paper Threaded Regions Schedulers Coverage
This PR aims to introduce 100% coverage of the [Paper Threaded Regions Schedulers](https://jd.papermc.io/paper/1.20/io/papermc/paper/threadedregions/scheduler/package-summary.html) and write its implementation in the same way as described on their Javadocs, with the only difference being that Bukkit is used as fallback scheduler when Folia isn't being used as server software.

This might seem redundant or even unnecessary, as at this stage one might be better off adding the [Folia API](https://jd.papermc.io/folia/1.20/) to its project dependencies, however, Paper forces everyone to use Java 17 when working with their API's which is just simply not feasible for plugins that just want to support every single version, including those being based upon Java 8.

## Reflection Performance
In comparison to the previous util, I cache as much of the reflections being performed as possible, and thus only the actual invocation of the method and the entity scheduler (I have to perform these reflections during runtime because I need the provided entity to retrieve the correct scheduler) has some overhead compared to when this is being done natively. Therefore it shouldn't be much of an issue, and those who aren't making use of this compatibility util won't be affected by this anyway.

## TaskWrapper
Since the threaded region schedulers of Paper are returning a [ScheduledTask](https://jd.papermc.io/paper/1.20/io/papermc/paper/threadedregions/scheduler/ScheduledTask.html) instead of a [BukkitTask](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/scheduler/BukkitTask.html), I introduced a general task wrapper, that adds support for methods supported by both the ScheduledTask as the BukkitTask, rather than nothing, as was being done previously. I am not too sure how I could further improve the TaskWrapper, so if someone has suggestions, I'll be happy to implement them!

## Deprecation
The old utility class (`io.github.retrooper.packetevents.util.FoliaCompatUtil`) has been marked as deprecated both through annotations and JavaDocs. In the JavaDocs I linked the new utility class (`io.github.retrooper.packetevents.util.folia.FoliaScheduler`), and where possible the specific corresponding scheduler/method.

## Remarks
The `FoliaScheduler` acts as the main entry to do anything related to Folia. I made the scheduler constructors protected, to prevent them from being directly initialized. I want to prevent people from unnecessarily creating extra instances of the schedulers as when wrongly done, would mean the reflections will be re-executed. I also added copyright notices on all the newly created files.

The `AsyncScheduler` uses a TimeUnit to determine the given delay, so when Folia isn't used we use the `AsyncScheduler#convertTimeToTicks` method to calculate the delay back to Minecraft ticks, as we need that value when invoking the Bukkit scheduler. 

### Showcase
```java
FoliaScheduler.getAsyncScheduler().runNow(plugin, (o) - > {
    // This will be executed asynchronously through the AsyncScheduler when Folia is used,
    // and otherwise it will fall back to Bukkit.getScheduler().runTaskAsynchronously()
});

FoliaScheduler.getRegionScheduler().runAtFixedRate(plugin, world, (o) - > {
    // This will be executed on the region thread specifically to the provided world after an initial delay of 20 ticks, and 
    // afterward every 50 ticks of the region, which owns the world
    // when Folia is used, and otherwise it will fall back
    // to Bukkit.getScheduler().runTaskTimer
}, 20, 50);
```

### CODE REVIEW

**Comment by Tofaa2** [comment]
W PR, the old util was horrible

**Review by Tofpu — CHANGES_REQUESTED** [review]

File: spigot/src/main/java/io/github/retrooper/packetevents/injector/handlers/PacketEventsDecoder.java [hunk 1]
```diff
@@ -86,7 +86,7 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
                 }
                 user.closeConnection();
                 if (player != null) {
-                    FoliaCompatUtil.runTaskForEntity(player, (Plugin) PacketEvents.getAPI().getPlugin(), () -> player.kickPlayer("Invalid packet"), null, 1);
+                    FoliaCompatUtil.getEntityScheduler().runDelayed(player, (Plugin) PacketEvents.getAPI().getPlugin(), (o) -> player.kickPlayer("PacketEvents 2.0 failed to inject"), null, 1);
```
[the remark this answers is not in the document]
Comment (Bram1903) [reply spigot/src/main/java/io/github/retrooper/packetevents/injector/handlers/PacketEventsDecoder.java:89]: Whoops my bad! 😅 Will fix once home!

[the remark this answers is not in the document]
Comment (Bram1903) [reply spigot/src/main/java/io/github/retrooper/packetevents/injector/handlers/PacketEventsDecoder.java]: Fixed 😄

**Comment by retrooper** [comment]
Thanks for the PR

**Pull request merged**